src/persistent-connections/content.md

```diff
@@ -0,0 +1,465 @@
+
+1. retrieve a list of endpoints from the Service
+1. for each of them, open a connection and keep it open
+1. when you need to make a request, pick one of the open connections
+1. on a regular interval refresh the list of endpoints and remove or add new connections
+
+```slideshow
+{
+  "description": "Client side load balancing",
+  "slides": [
+    {
+      "image": "assets/client-lb-1.svg",
+      "description": "Instead of having the red Pod issuing a request to your Service, you could load balance the request client-side."
+    },
+    {
+      "image": "assets/client-lb-2.svg",
+      "description": "Your could write some code that asks what Pods are part of the Service."
+    },
+    {
+      "image": "assets/client-lb-3.svg",
+      "description": "Once you have that list, you could store it locally and use it to connect to the Pods."
+    },
+    {
+      "image": "assets/client-lb-4.svg",
+      "description": "You are in charge of the load balancing algorithm."
+    }
+  ]
+}
+```
+
+_Does this problem apply only to keep-alive?_
+
+## Database persistent connections
+
+HTTP isn't the only protocol designed to open a long-lived connection.
+
+If your app uses a database, the connection isn't opened and closed every time you wish to retrieve a record or a document.
+
+Instead, the connection is established once and kept open.
+
+If your database is deployed in Kubernetes using a Service, you might experience the same issues as the previous example.
+
+There's one replica in your database that is utilised more than the others.
+
+Kube-proxy and Kubernetes don't help to balance persistent connections.
+
+Instead, you should take care of load balancing the requests to your database.
+
+Depending on the library that you use to connect to the database, you might have different options.
+
+The following example is from a clustered MySQL database called from Node.js:
+
+```js|title=index.js
+var mysql = require('mysql');
+var poolCluster = mysql.createPoolCluster();
+
+var endpoints = /* retrieve endpoints from the Service */
+
+for (var [index, endpoint] of endpoints) {
+  poolCluster.add(`mysql-replica-${index}`, endpoint);
+}
+
+// Make queries to the clustered MySQL database
+```
+
+As you can imagine, several other protocols open a single connection at the time and reuse it.
+
+Here you can read a few examples:
+
+- Websockets and secured WebSockets
+- HTTP/2
+- gRPC
+- RSockets
+- AMQP
+
+You might recognise most of the protocol above.
+
+_So if those protocols are so popular, why isn't there a standard answer to do load balancing rather than moving the logic into the client?_
+
+_Is there a native solution in Kubernetes?_
+
+Kube-proxy and iptables are designed to cover the most popular use cases of deployments in a Kubernetes cluster.
+
+But they are mostly there for convenience.
+
+If you're using a web service that exposes a REST API, then you're in luck — you can use any Kubernetes Service.
```

Comment:
If you're using a web service that exposes a REST API, then you're in luck — **this use case usually doesn't reuse TCP connections**, and you can use any Kubernetes Service.